example/mapreduce/MRBootstrap.go

```diff
@@ -0,0 +1,149 @@
+package mapreduce
+
+// user_interface provide user a interface to start their mapreduce application
+// MapreduceBootstrapController would controll the bootstrap of controller, task
+// (serves as same as regression/demo/run_regression.sh).
+// It likes a local version of Kubernetes, managing the whole framework status.
+// For user starting a new mapreduce framework, only need to do is
+// invoke NewMapreduceBootstrapController() to get a new controlle,
+// invoke its Start interface with his own mapreduce configuration(represents as map[string]interface{})
+// therefore the mapreduce framework will run automatically.
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"os"
+
+	"github.com/coreos/go-etcd/etcd"
+	"github.com/taskgraph/taskgraph"
+	"github.com/taskgraph/taskgraph/controller"
+	"github.com/taskgraph/taskgraph/framework"
+)
+
+type MRBootstrap struct {
+	Config      map[string]interface{}
+	logger      *log.Logger
+	taskBuilder taskgraph.TaskBuilder
+	topology    taskgraph.Topology
+}
+
+func NewMRBootstrap(taskBuilder taskgraph.TaskBuilder, topology taskgraph.Topology, userConfig map[string]interface{}) NewMRBootstrap {
+	return &NewMRBootstrap{
+		taskBuilder: taskBuilder,
+		topology:    topology,
+		Config:      userConfig,
+	}
+}
+
+const defaultBufferSize = 4096
+
+var controllerStarted chan bool
+var ntask uint64
+
+func (mpc *MRBootstrap) Start() error {
+
+	mpc.Config = config
+
+	mpc.checkConfiguration()
+
+	// calculate the maximum number of node coexist during all epochs
+	// plus one represents that thers is a reservation serving for master node
+	ntask = max(config["MapperNum"]+config["ShuffleNum"], config["ShuffleNum"]+config["ReducerNum"]) + 1
+	controllerStarted = make(chan bool, 1)
+
+	// Issue : could controller create a free work?
+	// I've not fingure out current framework how to create free work
+	// at my previous code, I directly add some path (Like FreeworkDir) the layout
+	// and add some value to this path
+	go mpc.runController(ntask)
+
+	// wait controller initialization finished
+	<-controllerStarted
+
+	mpc.logger = log.New(os.Stdout, "", log.Ldate|log.Ltime|log.Lshortfile)
+
+	for i := uint64(0); i < ntask+mpc.config["FreeNode"].(uint64); i++ {
```

Comment:
I don't know the purpose of this.
Is it a MR program for running on real machine? Or is it just for testing, demo?
If it's the second case, this is fine.